File: crates/database/src/entities/tournament.rs
use chrono::{
  DateTime, Utc,
  serde::{ts_seconds, ts_seconds_option},
};
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(
  Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize, EnumIter, DeriveActiveEnum,
)]
#[sea_orm(rs_type = "String", db_type = "String(StringLen::N(24))")]
#[serde(rename_all = "snake_case")]
pub enum Lifecycle {
  #[default]
  #[sea_orm(string_value = "draft")]
  Draft,
  #[sea_orm(string_value = "registration")]
  Registration,
  #[sea_orm(string_value = "running")]
  Running,
  #[sea_orm(string_value = "review")]
  Review,
  #[sea_orm(string_value = "finished")]
  Finished,
  #[sea_orm(string_value = "archived")]
  Archived,
}

impl Lifecycle {
  pub fn can_transition_to(self, next: Self) -> bool {
    matches!(
      (self, next),
      (Self::Draft, Self::Registration)
        | (Self::Registration, Self::Running)
        | (Self::Running, Self::Review)
        | (Self::Review, Self::Finished)
        | (Self::Finished, Self::Archived)
    )
  }

  pub fn locks_teams(self) -> bool {
    matches!(
      self,
      Self::Running | Self::Review | Self::Finished | Self::Archived
    )
  }
}

#[derive(
  Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize, EnumIter, DeriveActiveEnum,
)]
#[sea_orm(rs_type = "String", db_type = "String(StringLen::N(16))")]
#[serde(rename_all = "snake_case")]
pub enum CompetitionMode {
  #[default]
  #[sea_orm(string_value = "individual")]
  Individual,
  #[sea_orm(string_value = "team")]
  Team,
  #[sea_orm(string_value = "both")]
  Both,
}

#[derive(
  Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize, EnumIter, DeriveActiveEnum,
)]
#[sea_orm(rs_type = "String", db_type = "String(StringLen::N(16))")]
#[serde(rename_all = "snake_case")]
pub enum EvidencePolicy {
  #[default]
  #[sea_orm(string_value = "optional")]
  Optional,
  #[sea_orm(string_value = "required")]
  Required,
  #[sea_orm(string_value = "disabled")]
  Disabled,
}

#[derive(
  Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize, EnumIter, DeriveActiveEnum,
)]
#[sea_orm(rs_type = "String", db_type = "String(StringLen::N(16))")]
#[serde(rename_all = "snake_case")]
pub enum LeaderboardVisibility {
  #[default]
  #[sea_orm(string_value = "live")]
  Live,
  #[sea_orm(string_value = "frozen")]
  Frozen,
  #[sea_orm(string_value = "after_end")]
  AfterEnd,
}

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "tournament")]
pub struct Model {
  #[sea_orm(primary_key)]
  pub id: i64,
  pub name: String,
  pub brief: String,
  #[sea_orm(column_type = "Text", nullable)]
  pub description: Option<String>,
  pub owner_id: i64,
  pub lifecycle: Lifecycle,
  pub competition_mode: CompetitionMode,
  pub evidence_policy: EvidencePolicy,
  pub leaderboard_visibility: LeaderboardVisibility,
  pub cover: Option<String>,
  pub team_size_min: i32,
  pub team_size_max: i32,
  #[serde(with = "ts_seconds_option")]
  pub registration_start_at: Option<DateTime<Utc>>,
  #[serde(with = "ts_seconds_option")]
  pub registration_end_at: Option<DateTime<Utc>>,
  #[serde(with = "ts_seconds_option")]
  pub start_at: Option<DateTime<Utc>>,
  #[serde(with = "ts_seconds_option")]
  pub end_at: Option<DateTime<Utc>>,
  #[serde(with = "ts_seconds")]
  pub created_at: DateTime<Utc>,
  #[serde(with = "ts_seconds")]
  pub updated_at: DateTime<Utc>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
impl ActiveModelBehavior for ActiveModel {}

#[cfg(test)]
mod tests {
  use super::Lifecycle;

  #[test]
  fn lifecycle_only_moves_forward_one_step() {
    assert!(Lifecycle::Draft.can_transition_to(Lifecycle::Registration));
    assert!(Lifecycle::Review.can_transition_to(Lifecycle::Finished));
    assert!(!Lifecycle::Draft.can_transition_to(Lifecycle::Running));
    assert!(!Lifecycle::Finished.can_transition_to(Lifecycle::Review));
  }
}

File: web/src/lib/models/tournament.ts
import type { DateTime } from "luxon";

export type TournamentLifecycle = "draft" | "registration" | "running" | "review" | "finished" | "archived";
export type LifecycleScheduleMode = "manual" | "scheduled";
export type RoundReleaseAudience = "public" | "participants" | "staff";
export type RoundReleaseTiming = "immediate" | "on_enter" | "on_end";
export type RoundEndMode = "on_next_round" | "at_time" | "manual";
export type CompetitionMode = "individual" | "team" | "both";
export type EvidencePolicy = "required" | "optional" | "disabled";
export type LeaderboardVisibility = "live" | "frozen" | "after_end";
export type ChartVisibility = "public" | "after_archive" | "private";
export type ChartSourceType = "personal" | "phigros" | "phira";

export interface Tournament {
  id: number;
  name: string;
  brief: string;
  description?: string;
  owner_id: number;
  lifecycle: TournamentLifecycle;
  competition_mode: CompetitionMode;
  evidence_policy: EvidencePolicy;
  leaderboard_visibility: LeaderboardVisibility;
  cover?: string;
  team_size_min: number;
  team_size_max: number;
  registration_start_at?: DateTime;
  registration_end_at?: DateTime;
  start_at?: DateTime;
  end_at?: DateTime;
  registration_schedule: LifecycleScheduleMode;
  registration_at?: DateTime;
  running_schedule: LifecycleScheduleMode;
  running_at?: DateTime;
  review_schedule: LifecycleScheduleMode;
  review_at?: DateTime;
  finished_schedule: LifecycleScheduleMode;
  finished_at?: DateTime;
  organizer_can_edit_archived: boolean;
  current_round_id?: number;
  round_control_mode: "manual_assisted";
  created_at: DateTime;
  updated_at: DateTime;
}

export interface TournamentRound {
  id: number;
  tournament_id: number;
  name: string;
  description?: string;
  order_index: number;
  start_at?: DateTime;
  end_at?: DateTime;
  release_audience: RoundReleaseAudience[];
  release_timing: RoundReleaseTiming;
  end_mode: RoundEndMode;
  release_at?: DateTime;
  started_at?: DateTime;
  ended_at?: DateTime;
  released_at?: DateTime;
  manually_released: boolean;
  manually_ended: boolean;
  release_version: number;
}

export interface TournamentRoundConflict {
  current_round?: number;
  suggested_round: number;
  reason: string;
  affected_charts: number;
}
export interface ChartTag {
  id: number;
  tournament_id: number;
  round_id: number;
  name: string;
  order_index: number;
}
export interface TournamentChart {
  id: number;
  tournament_id: number;
  round_id: number;
  tag_id: number;
  title: string;
  artist: string;
  charter: string;
  difficulty: string;
  level_constant: number;
  cover?: string;
  description?: string;
  order_index: number;
  weight_millionths: number;
  metadata: Record<string, unknown>;
}
export interface ChartLibrary {
  id: number;
  title: string;
  artist: string;
  charter: string;
  difficulty: string;
  level_constant: number;
  weight_millionths?: number;
  cover?: string;
  metadata: Record<string, unknown>;
  source?: string;
  source_type?: ChartSourceType | string;
  tournaments?: string;
}
export interface TournamentChartLibrary {
  link: {
    id: number;
    tournament_id: number;
    chart_library_id: number | null;
    visibility: ChartVisibility;
    round_id: number;
    tag_id: number;
    order_index: number;
    weight_millionths: number;
    description?: string;
    title?: string;
    artist?: string;
    charter?: string;
    difficulty?: string;
    level_constant?: number;
    cover?: string;
    metadata?: Record<string, unknown>;
  };
  chart: ChartLibrary;
}
export interface Registration {
  id: number;
  tournament_id: number;
  user_id: number;
  display_name: string;
  status: "pending" | "approved" | "rejected" | "withdrawn";
  created_at: DateTime;
  updated_at: DateTime;
}
export interface TournamentTeam {
  id: number;
  tournament_id: number;
  name: string;
  captain_registration_id: number;
  created_at: DateTime;
}
export interface TournamentStaff {
  id: number;
  tournament_id: number;
  user_id: number;
  role: "owner" | "organizer" | "judge";
  created_at: DateTime;
}
export interface TournamentResult {
  id: number;
  tournament_id: number;
  chart_id: number;
  registration_id: number;
  team_id_snapshot?: number;
  submitted_by: number;
  score: number;
  accuracy_millionths: number;
  max_combo: number;
  full_combo: boolean;
  all_perfect: boolean;
  judgments: Record<string, number>;
  metrics: Record<string, unknown>;
  played_at: DateTime;
  evidence?: string;
  status: "pending" | "approved" | "rejected" | "voided";
  replaces_result_id?: number;
  created_at: DateTime;
}
export interface LeaderboardEntry {
  id: number;
  name: string;
  rank: number;
  score: number;
  tie_breakers: number[];
  breakdown?: unknown;
}
export interface LeaderboardSnapshot {
  id: number;
  tournament_id: number;
  kind: "individual" | "team";
  entries: LeaderboardEntry[];
  stale: boolean;
  error?: string;
  computed_at: DateTime;
}
export interface ScoringScript {
  id: number;
  tournament_id: number;
  version: number;
  name: string;
  template_key: string;
  source: string;
  source_hash: string;
  created_by: number;
  created_at: DateTime;
  active: boolean;
  immutable: boolean;
}
